### PR TITLE
FAH uses ids using the same naming policy as rolloutPolicy

### PR DESCRIPTION
**Pull request by inlined**
Additionally added listRollouts and fixed listBuilds so that it paginates over multiple requests and preserves the list of unavailable regions (and handles them)

### CODE REVIEW

File: src/gcp/apphosting.ts [hunk 1]
```diff
@@ -334,6 +347,33 @@ export async function getBuild(
...
+  const res: ListBuildsResponse = {
+    builds: [],
+    unreachable: [],
+  };
+
+  do {
+    const queryParams: Record<string, string> = pageToken ? { pageToken } : {};
+    const int = await client.get<ListBuildsResponse>(name, { queryParams });
+    res.builds.splice(res.builds.length, 0, ...(int.body.builds || []));
+    res.unreachable?.splice(res.unreachable.length, 0, ...(int.body.unreachable || []));
```
Comment (bkendall) [inline src/gcp/apphosting.ts:369]: Is it cleaner to use `.push`?

```suggestion
    res.builds.push(...(int.body.builds || []));
    res.unreachable?.push(...(int.body.unreachable || []));
```

File: src/gcp/apphosting.ts [hunk 2]
```diff
@@ -389,11 +429,24 @@ export async function listRollouts(
...
+  const res: ListRolloutsResponse = {
+    rollouts: [],
+    unreachable: [],
+  };
+
+  do {
+    const queryParams: Record<string, string> = pageToken ? { pageToken } : {};
+    const int = await client.get<ListRolloutsResponse>(name, { queryParams });
+    res.rollouts.splice(res.rollouts.length, 0, ...(int.body.rollouts || []));
+    res.unreachable?.splice(res.unreachable.length, 0, ...(int.body.unreachable || []));
```
Comment (bkendall) [inline src/gcp/apphosting.ts:444]: Similar to above... is `push` more clear?

File: src/gcp/apphosting.ts [hunk 3]
```diff
@@ -334,6 +347,33 @@ export async function getBuild(
...
+/**
+ * List Builds by backend
+ */
+export async function listBuilds(
+  projectId: string,
+  location: string,
+  backendId: string,
+): Promise<ListBuildsResponse> {
+  const name = `projects/${projectId}/locations/${location}/backends/${backendId}/builds`;
+  let pageToken: string | undefined = undefined;
```
Comment (bkendall) [inline src/gcp/apphosting.ts:359]: Technically I don't think you need to specify `= undefined` - that can be dropped.

```suggestion
  let pageToken: string | undefined;
```

File: src/gcp/apphosting.ts [hunk 4]
```diff
@@ -460,3 +516,53 @@ export async function ensureApiEnabled(options: any): Promise<void> {
   const projectId = needProjectId(options);
   return await ensure(projectId, API_HOST, "frameworks", true);
 }
+
+/**
+ * Generates the next build ID to fit with the naming scheme of the backend API.
+ * @param counter Overrides the counter to use, avoiding an API call.
+ * @return
```
Comment (bkendall) [inline src/gcp/apphosting.ts:523]: I _think_ you can drop an empty `@return` line - though I'm not certain

File: src/gcp/apphosting.ts [hunk 5]
```diff
@@ -460,3 +516,53 @@ export async function ensureApiEnabled(options: any): Promise<void> {
...
+export async function getNextRolloutId(
+  projectId: string,
+  location: string,
+  backendId: string,
+  counter?: number,
+): Promise<string> {
+  const date = new Date();
+  const year = date.getUTCFullYear();
+  // Note: month is 0 based in JS
+  const month = String(date.getUTCMonth()).padStart(2, "0");
```
Comment (bkendall) [inline src/gcp/apphosting.ts:534]: So should you be adding 1?

File: src/gcp/apphosting.ts [hunk 6]
```diff
@@ -460,3 +516,53 @@ export async function ensureApiEnabled(options: any): Promise<void> {
...
+  const day = String(date.getUTCDay()).padStart(2, "0");
+
+  if (counter) {
+    return `build-${year}-${month}-${day}-${String(counter).padStart(3, "0")}`;
+  }
+
+  // Note: must use exports here so that listRollouts can be stubbed in tests.
+  const builds = await (exports as { listRollouts: typeof listRollouts }).listRollouts(
```
Comment (bkendall) [inline src/gcp/apphosting.ts:542]: Huh, but `exports.listRollouts` gives you type errors or something? That's unfortunate :(

File: src/test/gcp/apphosting.spec.ts [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    });
+
+    function idPrefix(date: Date): string {
+      const year = date.getUTCFullYear();
+      const month = String(date.getUTCDay()).padStart(2, "0");
+      const day = String(date.getUTCDay()).padStart(2, "0");
+      return `build-${year}-${month}-${day}`;
+    }
+
+    it("Should handle explicit counters", async () => {
```
Comment (bkendall) [inline src/test/gcp/apphosting.spec.ts:24]: These generally shouldn't be capitalized (though I'm not sure I have any style guidance to point to). When I see this, I read it as a sentence, and "it Should do this" looks abnormal (compared to "it should do this")

**Review by bkendall — APPROVED** [review]
style/cleanup comments, but looks fine to me

File: src/gcp/apphosting.ts (hunk 5 above)
Comment (inlined) [reply src/gcp/apphosting.ts:534]: 😱 I modified the code and dropped the +1. Good catch and thank god for comments.

**Pull request merged**